### PR TITLE
Enforce page size limit in Parquet 

### PR DESCRIPTION
**Pull request by malhotrashivam**
For string columns and columns which use codecs, we don't enfoce page size limits properly. This was found while debugging #3328. The PR fixes the this issue. Also, this PR has some code refactoring to remove code duplication and make the code more readable.

### CODE REVIEW

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (lbooker42) [inline extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:92]: Love the comments, thank you!

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (lbooker42) [inline extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:878]: We typically only initialize non-default values.  `hasMoreDataToBuffer` is `false` by default.

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (lbooker42) [inline extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:1111]: This name doesn't seem correct.  Isn't this a chunk index?  maybe `currentChunkIdx` is more clear?

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (lbooker42) [inline extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:1131]: Do you keep around a reference to the old `buffer`? If not, can reuse this array and just reset the value.
```suggestion
                Arrays.fill(buffer, null);
```

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (lbooker42) [inline extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:1142]: Do we need to throw an exception or can we write the page and continue? Also, consider the case where we have a single string >= targetPageSize, shouldn't we write an abnormal page size and continue writing?

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (lbooker42) [inline extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:1203]: The binaries generated from the codec also are variable length and would benefit from the page size tests (similar to String).

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (lbooker42) [inline extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:1146]: As you noted in our call, this termination  may cause truncation when a vector of strings is to be written  that exceeds the target page size.  We need to handle vectors/arrays of strings differently and be accept potential OOM and oversize pages.

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (lbooker42) [inline extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:649]: I think I would leave a comment here explaining why you must grab the entire vector (because it must be written in a single page, etc.) and there is a risk of OOM exceptions.

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (lbooker42) [inline extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:1109]: This is a generic class now, so `String` references should go away in favor of something generic.

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (lbooker42) [inline extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:1096]: This abstraction is excellent.

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (lbooker42) [inline extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:888]: Instead of implementing this for each class, you **could** make it a `default` method on the `TransferObject` 
class. `EncodedTransfer` is the only one that actually needs a custom implementation.

What you have is fine, however, and might be better.  default methods can get you in trouble sometimes and we probably need to think carefully about this function with any new implementation of this class.

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (malhotrashivam) [reply extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:1109]: Nice catch, thanks!

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (malhotrashivam) [reply extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:888]: I wanted to do that, but like you explained, defaults were making me uncomfortable because part of the class needs to be specialized for each type and some parts can be default. So I kept it like this to make it easier to read instead of splitting across base and derived types.

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (malhotrashivam) [reply extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:1097]: Thanks!

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (lbooker42) [inline extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:925]: I don't think we need TransferObject and BaseTransfer.  You can make TransferObject an abstract class instead of an interface and move this common functionality to it.

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (lbooker42) [inline extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:956]: I think you'll get better performance if you make this a bulk transfer function.  Instead of making `chunk.size()` virtual calls (with the virtualization overhead), you can make a single call and pass in the size.

**Review by lbooker42 — DISMISSED** [review]

**Review by lbooker42 — APPROVED** [review]

**Review by chipkent — APPROVED** [review]

**Review by lbooker42 — DISMISSED** [review]

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (chipkent) [inline extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:824]:
```suggestion
            // We don't know the length of strings untill we read the actual data. Therefore, we take a relaxed estimate
```

**Review by chipkent — CHANGES_REQUESTED** [review]

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (malhotrashivam) [reply extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java]: Thanks for the comment.

**Review by chipkent — DISMISSED** [review]
LGTM

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (rcaudy) [inline extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:116]: I recommend a clearer name:
`transferOnePageToBuffer`
And a re-write to the JavaDoc to reflect that this is meant to be used when writing paginated data, and not meant to be interleaved with calls to the other method.

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (rcaudy) [inline extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:106]: I recommend a clearer name:
transferAllToBuffer
And a re-write to the JavaDoc to reflect that this is meant to be used when writing unpaginated data, and not meant to be interleaved with calls to the other method.

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (malhotrashivam) [reply extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:653]: I didn't paginate vector data because the implementation was becoming unnecessarily complicated in cases where we might run out of our page limit in the middle of the vector. 

- One approach is to pull out all elements of the vector, find the length in binary and then add it up to check if we are hitting the limit and only then transfer to buffer.
- Other way is to consider the limit in a relaxed manner and always put all elements of a single vector, and if we have hit or exceeded the limit, then stop.

Let me know if you want me to add this logic.

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (malhotrashivam) [reply extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:1119]: I agree with the >= comment.

For the optimisitic encoding, I tried that approach (it might be in one of the old commits) but then I was looking at how pyarrow does it and they are even more relaxed with the limit. They put elements in chunks of 100 in a page and then check if they have exceeded the limit. So I thought of keeping the implementation simpler by checking after buffering.
I can revert to the checking before buffering approach if you think that would be more appropriate.

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (malhotrashivam) [inline extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:92]: Once we complete the review and before merging, I am thinking of reorganizing this file by moving `TransferObject` closer to its child classes and reordering all methods in child classes to be in same order as this interface.

Comment (rcaudy) [reply extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:92]: Sounds great. Do it as a single, isolated commit, please.

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (rcaudy) [reply extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java]: I'm happy enough if the assumption is just declared and documented.

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (rcaudy) [reply extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java]: There's complexity at the layer above, too.
See `io.deephaven.parquet.table.pagestore.topage.ToVectorPage`
and `io.deephaven.parquet.table.pagestore.topage.ToArrayPage`

Comment (rcaudy) [reply extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java]: I think all we really want here is a comment that explains the problem (our code expecting all elements from an array or vector being on the same page, and a reference to that code) with a TODO referencing a new ticket that goes on the backlog.

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (rcaudy) [reply extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:92]: Try not to do anything that will introduce even more conflicts with https://github.com/deephaven/deephaven-core/pull/4479

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (rcaudy) [inline extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:967]: Prefer `io.deephaven.base.verify.Assert#eq(java.lang.Object, java.lang.String, java.lang.Object, java.lang.String)`

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (rcaudy) [inline extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:961]:
```suggestion
    private static final class IntTransfer extends PrimitiveTransfer<WritableIntChunk<Values>, IntBuffer> {
```

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (rcaudy) [inline extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:1159]: `Assert.neqNull` and `Assert.eqZero` or `Assert.assertion`

File: py/server/tests/test_parquet.py
Comment (rcaudy) [inline py/server/tests/test_parquet.py:289]: Ask @jmao-denver why this doesn't work and to approve as a Python code owner.

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (rcaudy) [reply extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:1193]: I was just suggesting to set `chunk == null` in the other method on this class.

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetInstructions.java
Comment (rcaudy) [inline extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetInstructions.java:449]: I see why you did this, but it's a little unfortunate. Is the default inappropriately high? Should we make it configurable?

File: extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java
Comment (rcaudy) [inline extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:105]: "long" -> "too big to fit in the available heap"?

Comment (malhotrashivam) [reply extensions/parquet/table/src/main/java/io/deephaven/parquet/table/ParquetTableWriter.java:105]: Thank you, this is much better :)

File: extensions/parquet/table/src/test/java/io/deephaven/parquet/table/ParquetTableReadWriteTest.java
Comment (rcaudy) [inline extensions/parquet/table/src/test/java/io/deephaven/parquet/table/ParquetTableReadWriteTest.java:93]: This isn't going to work reliably. If `ParquetInstructions` has been statically initialized, this will be ineffective. Why not just amend the tests below to set `MIN_TARGET_PAGE_SIZE` as the default, and try to write pages bigger than that?

Comment (malhotrashivam) [reply extensions/parquet/table/src/test/java/io/deephaven/parquet/table/ParquetTableReadWriteTest.java:93]: Will do!

File: extensions/parquet/table/src/test/java/io/deephaven/parquet/table/ParquetTableReadWriteTest.java
Comment (rcaudy) [inline extensions/parquet/table/src/test/java/io/deephaven/parquet/table/ParquetTableReadWriteTest.java:856]: I might have said to explicitly reference the constant, but this is fine.

**Review by rcaudy — APPROVED** [review]

File: extensions/parquet/table/src/test/java/io/deephaven/parquet/table/ParquetTableReadWriteTest.java
Comment (malhotrashivam) [reply extensions/parquet/table/src/test/java/io/deephaven/parquet/table/ParquetTableReadWriteTest.java:856]: Will make the change in another PR

**Pull request merged**
